src/doc_table.c

```diff
@@ -9,24 +9,76 @@
 #include "rmalloc.h"
 #include "spec.h"
 
+#define MAX_SIZE 1000000
+
 /* Creates a new DocTable with a given capacity */
 DocTable NewDocTable(size_t cap) {
   return (DocTable){.size = 1,
                     .cap = cap,
                     .maxDocId = 0,
                     .memsize = 0,
                     .sortablesSize = 0,
-                    .docs = rm_calloc(cap, sizeof(RSDocumentMetadata)),
+                    .maxSize = MAX_SIZE,
+                    .buckets = rm_calloc(cap, sizeof(DMDChain)),
                     .dim = NewDocIdMap()};
 }
 
+static inline uint32_t DocTable_GetBucket(DocTable *t, t_docId docId) {
+  return docId < t->maxSize ? docId : docId % t->maxSize;
+}
+
+static inline int DocTable_ValidateDocId(DocTable *t, t_docId docId) {
+  return docId != 0 && docId <= t->maxDocId;
+}
+
 /* Get the metadata for a doc Id from the DocTable.
  *  If docId is not inside the table, we return NULL */
+
+int DMDChain_IsEmpty(DMDChain *dmdChain) {
+  return !dmdChain->last;
+}
+
 inline RSDocumentMetadata *DocTable_Get(DocTable *t, t_docId docId) {
-  if (docId == 0 || docId > t->maxDocId) {
+  if (!DocTable_ValidateDocId(t, docId)) {
     return NULL;
   }
-  return &t->docs[docId];
+  uint32_t bucketIndex = DocTable_GetBucket(t, docId);
+  DMDChain *dmdChain = &t->buckets[bucketIndex];
+  RSDocumentMetadata *currDmd = dmdChain->first;
+  while (currDmd) {
+    if (currDmd->id == docId) {
+      return currDmd;
+    }
+    currDmd = currDmd->next;
+  }
+  return NULL;
+}
+
+static inline void DocTable_Set(DocTable *t, t_docId docId, RSDocumentMetadata dmd) {
+  uint32_t bucket = DocTable_GetBucket(t, docId);
+  if (bucket >= t->cap && t->cap < t->maxSize) {
+    size_t oldcap = t->cap;
+    t->cap += 1 + (t->cap ? MIN(t->cap / 2, 1024 * 1024) : 1);
```

Comment:
can you please comment what this does? I mean, it's obviously some kind of growing algorithm, but what exactly does it do.

Reply:
Also, if you grow the array, don't you need to rebalance all the chains? or is that why you only do modulo when the capacity is exceeded.